From 72ff2a2c90e75a9be85f31066fcb1fd89530f688 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 31 Oct 2024 12:16:50 +0200
Subject: [PATCH] [FSW-2071] boot_desc: get descriptor by index

Until now there was an option to get image descriptor by type, but
in order to be able to iterate over image descriptors an api is added
to get image descriptors by index.

Change-Id: I6dfb743b91ded98908797f3bc4e359409f97b97f
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/include/boot_desc.h b/include/boot_desc.h
index ee189fa..6ad8792 100644
--- a/include/boot_desc.h
+++ b/include/boot_desc.h
@@ -54,6 +54,7 @@
  */
 u32 boot_desc_get_img_desc(enum fw_post_reset_cmd upd_status,
 			   enum hl_fw_img_type type, struct boot_desc *desc);
+u32 boot_desc_get_img_desc_by_idx(unsigned int idx, struct boot_desc *desc);
 
 bool is_legacy_boot_descs(struct hl_fw_boot_ptrs *boot_ptrs);
 bool dev_spec_img_desc_is_valid(struct hl_fw_img_desc *img_desc);
diff --git a/lib/common/boot_desc.c b/lib/common/boot_desc.c
index 1ae0341..ee3a3d6 100644
--- a/lib/common/boot_desc.c
+++ b/lib/common/boot_desc.c
@@ -194,3 +194,15 @@
 	preboot_puts(" - not found!\n");
 	return HL_FW_IMG_NOT_FOUND;
 }
+
+u32 boot_desc_get_img_desc_by_idx(unsigned int idx, struct boot_desc *desc)
+{
+	struct hl_fw_img_desc *desc_ptr;
+
+	if (idx >= boot_ptrs.count)
+		return HL_FW_IMG_NOT_FOUND;
+
+	desc_ptr = (struct hl_fw_img_desc *)(desc_buf + idx * boot_ptrs.img_desc_sz);
+
+	return dev_spec_get_img_desc(desc_ptr, FLOPS_UPD_NOOP, desc);
+}
